funcs/generalFunctions.py

```diff
@@ -6,6 +6,60 @@
 import numpy as np
 import json
 
+
+def getLepIdxFrom4Vec(entry, lep_4vec, lep_type):
+    """ Classification: genMatching
+           - this function returns the index of a 4-vector (lep_4vec) belonging to
+             the muon (lep_type = 'm') or electron (lep_type = 'e') collection
+    """
+    
+    idx_lep = -1
+    if lep_type == 'm':
+        for i in range(entry.nMuon):
+            if abs((lep_4vec.Pt() - entry.Muon_pt[i])/lep_4vec.Pt()) < 0.08:
+                if abs((lep_4vec.Eta() - entry.Muon_eta[i])/lep_4vec.Eta()) < 0.08:
+                    if abs((lep_4vec.Phi() - entry.Muon_phi[i])/lep_4vec.Phi()) < 0.08:
+                        idx_lep = i
+
+    if lep_type == 'e':
+        for j in range(entry.nElectron):
+            if abs((lep_4vec.Pt() - entry.Electron_pt[j])/lep_4vec.Pt()) < 0.08:
+                    if abs((lep_4vec.Eta() - entry.Electron_eta[j])/lep_4vec.Eta()) < 0.08:
+                        if abs((lep_4vec.Phi() - entry.Electron_phi[j])/lep_4vec.Phi()) < 0.08:
```

Comment:
what the 0.08 stands for ? and has to be really hardcoded ?

Reply:
This function just takes a 4-vector and finds the index of the particle corresponding to it. So the 0.08 means similarity within 8% between pt, phi, and eta. It's arbitrary, but much easier algorithmically than finding a "closest match" particle.